### PR TITLE
Add 1D/2D Spatial Recall experiments, Causal Hyena, and infrastructure improvements

### PR DESCRIPTION
**Pull request by dwromero**
## Summary

This PR introduces **comprehensive 1D and 2D Spatial Recall experiments** for benchmarking sequence models (Hyena, Mamba, Attention) on dense prediction tasks, along with significant infrastructure improvements including **Causal Hyena support**, autoregressive training, checkpointing, config validation, and experiment tracking.

**Total: 154 files changed, 16,143 insertions, 500+ experiments completed**

---

### 1D Spatial Recall - Hyena L_cache=64 Discovery

**Key Finding**: Hyena with `L_cache=64` achieves **~1900x better** performance than Attention on 1D spatial recall!

| Model | Best Val Loss | Config | Notes |
|-------|--------------|--------|-------|
| **Hyena XS (L_cache=64)** | **0.000047** 🏆 | readout=-1, 22k steps | BEST! Breakthrough result |
| **Mamba M (patchify p=8)** | **0.0004** | readout=-1, 120k steps | Extended training + marker |
| **Hyena (patchify p=4)** | **0.00007** | 1024 tokens, 20k steps | Best patchified |
| Attention XS (rope=10) | 0.090 | 20k steps | Best rope_base, still ~1900x worse |

**What is L_cache?** The SIREN kernel's positional grid resolution. `L_cache=64` means coarser grid (step_size=1/63 vs 1/4095 for L_cache=4096).

#### Simple Copy Task (Best at 20k iterations)

| Architecture | Best Val Loss | Config | Notes |
|--------------|--------------|--------|-------|
| **Hyena M** | **0.000057** 🏆 | No patchify | Best overall |
| **Mamba M + Patch** | 0.0009 (p=4) | Patchify | 610× improvement with patchification! |
| **Attention M + Patch** | 0.000132 (p=32) | Patchify | Requires large patches |

#### Mask Selection Task (100k iterations)

| Architecture | Best Val Loss | Config | Notes |
|--------------|--------------|--------|-------|
| **Hyena M (patchify p=2)** | **0.00129** 🏆 | 100k iter | 17× better than 20k! |
| **Mamba M (patchify p=4)** | **0.00295** | 100k iter | 15× improvement |
| **Attention M (patchify p=8)** | **0.0161** | 100k iter | 6× improvement |

#### Color Selection Task (100k iterations)

| Architecture | Best Val Loss | Config | Notes |
|--------------|--------------|--------|-------|
| **Mamba M (patchify p=4)** | **0.000415** 🏆 | 100k iter | Best with marker! |
| **Hyena M (none)** | **0.000517** 🥈 | 100k iter | Non-patchify works! |
| **Hyena S (patchify p=2)** | **0.000588** 🥉 | 100k iter | |

#### Color Conditioning Task (50k iterations)

| Architecture | Best Val Loss | Config | Notes |
|--------------|--------------|--------|-------|
| **Hyena M (none)** | **0.0028** 🏆 | 50k iter | Non-patchify best! |
| **Hyena M (patchify p=2)** | 0.0043 | 50k iter | |
| **Mamba M (patchify p=4)** | 0.0050 | 50k iter | |

---

### Task Description

1D version of spatial recall where:
- Images are **flattened FIRST** (16×16 → 256 elements)
- Flattened image placed as contiguous segment in 1D canvas (4096 elements)
- Models are **causal** (can only see past, not future)

### Key Findings

1. **Hyena L_cache=64 dominates** - ~1900x better than Attention
2. **Mamba with readout=-1.0 marker dramatically improves** - 4.6x better at 1D simple copy
3. **Extended training (120k) + patchify + marker**: Mamba achieves 0.0004 val loss!
4. **Attention rope_base sweep**: Lower values help (10→0.090 best), but still ~1900x worse

### Experiment Variants

- **Simple Copy**: Single item, fixed position
- **Mask Selection**: 4 items with binary mask indicating target

---

### Task Variants

| Task | Input | Output | Selection |
|------|-------|--------|-----------|
| **Simple Copy** | 1ch grayscale (64×64) | 16×16 target | Fixed position |
| **Mask Selection** | 2ch (grayscale + mask) | 16×16 target | Binary mask |
| **Color Selection** | 3ch RGB (colored frames) | 1ch grayscale | Colored bounding box |
| **Color Conditioning** | 3ch RGB (colored frames) | 3ch RGB target | Color + color memory |

### Key Findings

1. **Hyena wins without patchification** at sufficient model size (M)
2. **Mamba needs patchification** - 610× improvement with p=4!
3. **Extended training (5×) helps ALL architectures** - 6-17× improvements
4. **Optimal patch sizes**: Hyena p=2, Mamba p=4, Attention p=8-32
5. **Attention struggles** with color tasks regardless of configuration

---

### New Modules

- **`nvsubquadratic/modules/causal_conv1d.py`**: 1D convolution with causal (left-only) padding
- **`CKConvND.is_causal`**: Flag to enable causal FFT convolutions for 1D data

### Causality Tests

**New Files:**
- `tests/test_causality_hyena.py` - 21 tests for Hyena, CausalConv1D, CKConvND
- `tests/test_causality_attn_mamba.py` - 20 tests for Attention & Mamba

Comprehensive test suite verifying:
- Forward pass causality (future independence)
- Gradient flow (no gradients to future positions)
- Integration with QKVSequenceMixer

---

### 1. SpatialRecallDataset Enhancements

**File:** `experiments/datamodules/spatial_recall_dataset.py`

New features:
- **1D mode**: `SpatialRecall1DDataset` and `SpatialRecall1DDataModule`
- **Colored frames mode**: RGB canvas with colored bounding boxes
- **Color conditioning**: Output RGB labels colored with frame color
- **Multiple items**: Support for distractors on canvas
- **readout_value**: Explicit marker for readout region

### 2. Autoregressive Training Support

**New Files:**
- `experiments/lightning_wrappers/autoregressive_wrapper.py` - Lightning wrapper for next-token prediction
- `tests/test_autoregressive_wrapper.py` - Comprehensive unit tests (445 lines)

### 3. Start From Checkpoint Feature

**Files:** `experiments/run.py`, `experiments/default_cfg.py`

New `StartFromCheckpointConfig` to load pretrained weights without resuming full training state:
- Downloads checkpoint from W&B artifacts
- Supports `strict` and `partial_load` modes
- Callbacks for state_dict processing before loading

### 4. Autoresume from W&B Artifacts

**File:** `experiments/run.py`

Enhanced autoresume to download checkpoints from W&B:
- Searches W&B for existing runs by name
- Downloads checkpoint artifact if found
- Falls back to local `last.ckpt` if no W&B checkpoint
- Properly attaches to existing W&B run

### 5. Config Override Validation

**File:** `experiments/utils/cli.py`

Invalid config overrides now raise clear errors:
```
Invalid config override: 'train.iteratoins'. 
Key 'iteratoins' does not exist in config. 
Available keys: ['do', 'precision', 'iterations', ...]
```

### 6. 1D Sequence Visualization Callback

**New File:** `experiments/callbacks/sequence_visualization_1d.py`

### 7. Ruff Linting Configuration

**File:** `pyproject.toml`

Added `RUF001`, `RUF002`, `RUF003` to ignore list.

---

## Documentation

Comprehensive experiment tracking with W&B links:

- **[1D Simple Copy TRACKER.md](examples/spatial_recall_1d/emnist_regression_simple_copy/TRACKER.md)** - 660+ lines
- **[1D Mask Selection TRACKER.md](examples/spatial_recall_1d/emnist_regression_mask_selection/TRACKER.md)** - 660+ lines
- **[2D Simple Copy TRACKER.md](examples/spatial_recall_2d/emnist_regression_simple_copy/TRACKER.md)** - 410+ lines
- **[2D Mask Selection TRACKER.md](examples/spatial_recall_2d/emnist_regression_mask_selection/TRACKER.md)** - 280+ lines
- **[2D Color Selection TRACKER.md](examples/spatial_recall_2d/emnist_regression_color_selection/TRACKER.md)** - 315+ lines
- **[2D Color Conditioning TRACKER.md](examples/spatial_recall_2d/emnist_regression_color_conditioning/TRACKER.md)** - 226+ lines

---

## Visualizations

Added example visualizations for all task variants:

| Task | Visualization |
|------|---------------|
| 1D Simple Copy | `visualizations/spatial_recall/spatial_recall_emnist_1d_simple_copy.png` |
| 1D Mask Selection | `visualizations/spatial_recall/spatial_recall_emnist_1d_mask_selection_4items.png` |
| 2D Simple Copy | `visualizations/spatial_recall/spatial_recall_emnist_2d_simple_copy.png` |
| 2D Mask Selection | `visualizations/spatial_recall/spatial_recall_emnist_2d_mask_selection_4items.png` |
| 2D Color Selection | `visualizations/spatial_recall/spatial_recall_emnist_2d_color_selection_4items.png` |
| 2D Color Conditioning | `visualizations/spatial_recall/spatial_recall_emnist_2d_color_conditioning_4items.png` |

---

## Test Plan

- [x] All 1D experiments completed (300+ runs)
- [x] All 2D experiments completed (200+ runs)
- [x] Unit tests for AutoregressiveWrapper pass
- [x] Unit tests for causality (41 tests total) pass
- [x] Config validation tests pass
- [x] Causal Hyena tests pass (21 tests)
- [x] WandB logging verified
- [x] Start from checkpoint feature tested
- [x] Autoresume from W&B tested

---

### XS (Extra-Small) Models (~0.7-0.8M params)

| Model | Hidden Dim | Heads/Headdim | Params | Notes |
|-------|-----------|---------------|--------|-------|
| Attention | 160 | 8 heads (head_dim=20) | ~719K | Causal attention with RoPE |
| Mamba | 128 | headdim=32, expand=2 | ~738K | Unidirectional (bidirectional=False) |
| Hyena | 160 | SIREN kernel, 3 layers | ~757K | Causal CKConvND + CausalConv1D |

### S (Small) Models (~1.8-2.0M params)

| Model | Hidden Dim | Params |
|-------|-----------|--------|
| Attention | 256 | ~1.84M |
| Mamba | 160-224 | ~1.91M |
| Hyena | 256 | ~1.91M |

### M (Medium) Models (~4-5M params)

| Model | Hidden Dim | Params |
|-------|-----------|--------|
| Attention | 384 | ~4.4M |
| Mamba | 256-352 | ~4.5M |
| Hyena | 416 | ~4.97M |

### CODE REVIEW

**Comment by moradza** [comment]
@dwromero can you explain the task here through an example. Given image x what is the desired output and label.

**Comment by dwromero** [comment]
@moradza you can look at examples under visualizations. I'll update those images in a while so they are updated with all the different variations we have now. I also need to update the PR as many more exps are finished now.

**Comment by dwromero** [comment]
@moradza with the latest update, it should be clear what experiment is solving

**Review by Dafidofff — APPROVED** [review]
LGTM!!

**Review by moradza — APPROVED** [review]
LGTM!
